README.md
Angular Truncate
----------------

[![License](https://img.shields.io/npm/l/angular-truncate-2.svg)](http://bower.io/search/?q=angular-truncate-2)
[![Build Status](https://travis-ci.org/BernardoSilva/angular-truncate-2.svg?branch=master)](https://travis-ci.org/BernardoSilva/angular-truncate-2)
[![Coverage Status](https://coveralls.io/repos/BernardoSilva/angular-truncate-2/badge.svg)](https://coveralls.io/r/BernardoSilva/angular-truncate-2)
[![Bower](https://img.shields.io/bower/v/angular-truncate-2.svg)](http://bower.io/search/?q=angular-truncate-2)
[![NPM](https://img.shields.io/npm/v/angular-truncate-2.svg)](https://www.npmjs.com/package/angular-truncate-2)


This project is a filter for Angularjs to truncate text strings to a set number of characters or words and
add ellipses when needed.

[Demo](http://bernardosilva.github.io/angular-truncate-2/)

## Install

You can install this package with `bower`, `npm` and `gem`.

### bower

[View in Bower](http://bower.io/search/?q=angular-truncate-2)

```shell
bower install angular-truncate-2
```

The files are then available at `bower_components/angular-truncate-2/src/angular-truncate-2.js`

### npm

[View in NPM](https://www.npmjs.com/package/angular-truncate-2)

```shell
npm install angular-truncate-2
```

The files are then available at `node_modules/angular-truncate-2/src/angular-truncate-2.js`


### gem

Converted using [rails-asset](https://rails-assets.org/).

```shell
gem 'rails-assets-angular-truncate-2'
```
Include following in application.js:
```js
//= require angular-truncate-2
```

## How to use angular truncate


###Include the javascript file.

``` html
<script src="angular-truncate-2.min.js"></script>
```

###Inject the `truncate` filter into your app module.

```javascript
var myApp = angular.module('myApp', ['truncate']);
```

###When outputting text, apply the filter.
```html
 <p>
    {{ text | characters:25 }} or {{ text | words:5 }}
</p>
```

By default, a _word_ will not be truncated. Set the optional boolean after the character count to true.
```html
 <p>
     {{ text | characters:25 :true}}
 </p>
 ```

Filters can also be chained together. It will truncate after 25 words or before 25 words if the 100 character limit is reached.
 ```html
  <p>
     {{ text | words:25 | characters: 100}}
 </p>
 ```

## How contribute

* Fork the repository
* Do your changes/suggestions
* Create a Pull Request


### How to run tests

```sh
npm test
```


This project is based on [angular-truncate](https://github.com/sparkalow/angular-truncate) created by Brian Mathews.


CHANGELOG.md
# Angular-truncate changelog

### 0.4.2

- Remove dependencies from version control to bower in [695a7bf](https://github.com/BernardoSilva/angular-truncate-2/commit/695a7bf70a2583ff119a981f9fc8e44dd44f213a) by [osavchenko](https://github.com/osavchenko).

### 0.4.1

- Bump code coverage to 100%.
- Fix build badge path.
- Fix code coverage badge path.

### 0.4.0

- Renamed all files to match project name.

### 0.3.0 (2015-10-31)

- Add `postfix` param for words filter in [db16501](https://github.com/BernardoSilva/angular-truncate/commit/db1650133f37f045e9281089b11aa724474b211d)
- Add minified version.

### 0.2.0 (2015-05-25)

- Add code coverage. [PR 15](https://github.com/BernardoSilva/angular-truncate/pull/16)
- Fix bower install warnings about version and missing ignore entry. [PR 17](https://github.com/BernardoSilva/angular-truncate/pull/17)


test/karma.conf.js
// Testacular configuration file
//
// For all available config options and default values, see:
// https://github.com/vojtajina/testacular/blob/stable/lib/config.js#L54

var configuration = {
    // other things
 
    


	// base path, that will be used to resolve files and exclude
	basePath: './..',

	// frameworks to use
	frameworks: ['jasmine'],

	// list of files / patterns to load in the browser
	files: [
	  'test/lib/angular/angular.min.js',
	  'test/lib/angular/angular-mocks.js',
	  'src/*.js',
	  'test/unit/*.js'
	],

	// list of files to exclude
	exclude: [],

	// use dots reporter, as travis terminal does not support escaping sequences
	// possible values: 'dots', 'progress', 'junit'
	// CLI --reporters progress
	// 'coverage',
	reporters: ['progress'],

	// web server port
	// CLI --port 9876
	port: 9876,

	// cli runner port
	// CLI --runner-port 9100
	runnerPort: 9100,

	// enable / disable colors in the output (reporters and logs)
	// CLI --colors --no-colors
	colors: true,

	// level of logging
	// possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
	// CLI --log-level debug
	//logLevel: LOG_INFO,

	// enable / disable watching file and executing tests whenever any file changes
	// CLI --auto-watch --no-auto-watch
	autoWatch: true,

	// Start these browsers, currently available:
	// - Chrome
	// - ChromeCanary
	// - Firefox
	// - Opera
	// - Safari (only Mac)
	// - PhantomJS
	// - IE (only Windows)
	// CLI --browsers Chrome,Firefox,Safari
	browsers: ['Chrome'],

	// If browser does not capture in given timeout [ms], kill it
	// CLI --capture-timeout 5000
	captureTimeout: 5000,

	// Auto run tests on start (when browsers are captured) and exit
	// CLI --single-run --no-single-run
	singleRun: false,

	// report which specs are slower than 500ms
	// CLI --report-slower-than 500
	reportSlowerThan: 500,

	customLaunchers: {
        ChromeNoSandbox: {
        	base: 'Chrome',
        	flags: ['--no-sandbox'] 
    	}
    },

};
 
if (process.env.TRAVIS) {
    configuration.browsers = ['ChromeNoSandbox'];
}
 
module.exports = function (config) {

config.set(configuration);
};


test/filtersSpec.js
/*global module, inject, beforeEach, expect, describe, it */

'use strict';

describe('truncate', function () {

    beforeEach(module('truncate'));

    describe('characters', function () {
        var characterFilter;

        beforeEach(inject(function ($filter) {
            characterFilter = $filter('characters');
        }));

        it('should do nothing to this string', function () {
            expect(characterFilter('1234567890')).toEqual('1234567890');
        });

        it('should fail', function () {
            expect(characterFilter(null, 30)).not.toEqual('1234567890');
        });

        it('should not trim these down', function () {
            expect(characterFilter('1234567890', 30)).toEqual('1234567890');
        });

        it('should trim these down', function () {
            expect(characterFilter('1234567890', 5)).toEqual('12345…');
        });

        it('should trim this down including the space', function () {
            expect(characterFilter('123456789 10 11 12 13 14', 13)).toEqual('123456789 10…');
        });

        it('should trim this down breaking on words', function () {
            expect(characterFilter('123456789 10 11 12 13 14', 14, true)).toEqual('123456789 10 1…');
        });

        /**
         * This example the trim will happen on a space it should
         * trim to last chracter and add ellipses.
         */
        it('should trim this down breaking on words last character and ignore spaces', function () {
            expect(characterFilter('123456   789', 7, true)).toEqual('123456…');
        });

        it('should trim this down ignoring the space', function () {
            expect(characterFilter('Florida/New Jersey/California/Texas', 30, true)).toEqual('Florida/New Jersey/California/…');
        });

        it('should handle invalid numbers', function () {
            expect(characterFilter('1234567890', 0)).toEqual('');
        });

        it('should handle invalid chars numbers type', function () {
            expect(characterFilter('1234567890', 'abc')).toEqual('1234567890');
        });
    });

    describe('words', function () {
        var wordFilter;

        beforeEach(inject(function ($filter) {
            wordFilter = $filter('words');
        }));

        it('should do nothing to this string', function () {
            expect(wordFilter('1234567890')).toEqual('1234567890');
        });

        it('should do nothing to this multi words string', function () {
            expect(wordFilter('1234567890 abc def')).toEqual('1234567890 abc def');
        });

        it('should fail', function () {
            expect(wordFilter(null, 30)).not.toEqual('1234567890');
        });

        it('should not trim these down', function () {
            expect(wordFilter('1234567890', 1)).toEqual('1234567890');
        });

        it('should trim these down', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 5)).toEqual('abc def ghhi jkl mno…');
        });

        it('should trim these down and handle multi-spaces', function () {
            expect(wordFilter('abc def    ghhi jkl    mno pqr stu    vw   xyz', 5)).toEqual('abc def ghhi jkl mno…');
        });

        it('should not trim invalid words numbers', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 0)).toEqual('');
        });

        it('should not trim invalid words type', function () {
            expect(wordFilter('hello how u doin', 'abc')).toEqual('hello how u doin');
        });

        it('should not trim higher words numbers', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 25)).toEqual('abc def ghhi jkl mno pqr stu vw xyz');
        });

        it('should trim these down without ellipsis postfix', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 5, '')).toEqual('abc def ghhi jkl mno');
        });

        it('should trim these down with specific postfix', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 5, '...')).toEqual('abc def ghhi jkl mno...');
        });

        it('should trim these down with blank postfix', function () {
            expect(wordFilter('abc def ghhi jkl mno pqr stu vw xyz', 5, '')).toEqual('abc def ghhi jkl mno');
        });

    });

    describe('splitcharacters', function () {
        var characterFilter;

        beforeEach(inject(function ($filter) {
            characterFilter = $filter('splitcharacters');
        }));

        it('should do nothing to this string', function () {
            expect(characterFilter('1234567890')).toEqual('1234567890');
        });

        it('should fail', function () {
            expect(characterFilter(null, 30)).not.toEqual('1234567890');
        });

        it('should not trim these down', function () {
            expect(characterFilter('1234567890', 30)).toEqual('1234567890');
        });

        it('should trim these down', function () {
            expect(characterFilter('1234567890', 5)).toEqual('12...890');
        });

        it('should trim this down including the space', function () {
            expect(characterFilter('123456789 10 11 12 13 14', 13)).toEqual('123456...2 13 14');
        });

        it('should handle invalid numbers', function () {
            expect(characterFilter('1234567890', 0)).toEqual('');
        });

        it('should handle invalid chars numbers type', function () {
            expect(characterFilter('1234567890', 'abc')).toEqual('1234567890');
        });
    });
});


Gruntfile.js
/* global require, module, process */

'use strict';

module.exports = function(grunt) {

  // Project configuration.
  grunt.initConfig({
    pkg: grunt.file.readJSON('package.json'),
    meta: {
      banner: '/**\n' +
      ' * <%= pkg.description %>\n' +
      ' * @version v<%= pkg.version %> - <%= grunt.template.today("yyyy-mm-dd") %>\n' +
      ' * @link <%= pkg.homepage %>\n' +
      ' * @author <%= pkg.author %>\n' +
      ' * @license MIT License, http://www.opensource.org/licenses/MIT\n' +
      ' */\n'
    },
    dirs: {
      dest: 'dist'
    },
    concat: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        src: ['src/*.js'],
        dest: '<%= dirs.dest %>/<%= pkg.name %>.js'
      }
    },
    uglify: {
      options: {
        banner: '<%= meta.banner %>'
      },
      dist: {
        src: ['<%= concat.dist.dest %>'],
        dest: '<%= dirs.dest %>/<%= pkg.name %>.min.js'
      }
    },
    jshint: {
      files: ['Gruntfile.js', 'src/*.js', 'test/unit/*.js'],
      options: {
        curly: false,
        browser: true,
        eqeqeq: true,
        immed: true,
        latedef: true,
        newcap: true,
        noarg: true,
        sub: true,
        undef: true,
        boss: true,
        eqnull: true,
        expr: true,
        node: true,
        globals: {
          exports: true,
          angular: false,
          $: false
        }
      }
    },
    // watch: {
    //   files: '<config:jshint.files>',
    //   tasks: 'default'
    // },
    karma: {
      test: {
        options: {
          reporters: ['dots'],
          singleRun: true
        }


      },
      options: {
        configFile: 'test/karma.conf.js'
      }
    }
  });

  // Load the plugin that provides the "jshint" task.
  grunt.loadNpmTasks('grunt-contrib-jshint');

  // Load the plugin that provides the "concat" task.
  grunt.loadNpmTasks('grunt-contrib-concat');

  // Load the plugin that provides the "uglify" task.
  grunt.loadNpmTasks('grunt-contrib-uglify');

  grunt.loadNpmTasks('grunt-contrib-connect');

  // Load the plugin that provides the "watch" task.
  //grunt.loadNpmTasks('grunt-contrib-watch');


  // Default task.
  grunt.registerTask('default', ['test']);

  // Test tasks.
  grunt.registerTask('test', ['jshint', 'karma:test']);
  grunt.registerTask('test-server', ['karma:server']);

  // Build task, `test` was removed because it's not stable.
  grunt.registerTask('build', ['concat', 'uglify']);

  // run devserver
  grunt.registerTask('webserver', ['connect:devserver']);

  // Provides the "karma" task.
  grunt.registerMultiTask('karma', 'Starts up a karma server.', function() {
    var done = this.async();
    require('karma').server.start(this.options(), function(code) {
      done(code === 0);
    });
  });

};


karma.conf.js
// Testacular configuration file
//
// For all available config options and default values, see:
// https://github.com/vojtajina/testacular/blob/stable/lib/config.js#L54

var configuration = {
    // other things


    // base path, that will be used to resolve files and exclude
    // basePath: './..',

    // frameworks to use
    frameworks: ['jasmine'],

    // list of files / patterns to load in the browser
    files: [
        'bower_components/angular/angular.js',
        'bower_components/angular-mocks/angular-mocks.js',
        'src/*.js',
        'test/*.js'
    ],

    // list of files to exclude
    exclude: [],

    preprocessors: {
        'src/*.js': ['coverage']
    },

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress', 'junit'
    // CLI --reporters progress
    // 'coverage',
    reporters: ['progress', 'coverage', 'dots'],

    coverageReporter: {
        // specify a common output directory
        dir: 'build/reports/coverage',
        reporters: [
            // reporters not supporting the `file` property
            {type: 'html', subdir: 'report-html'},
            {type: 'lcov', subdir: 'report-lcov'},
            // reporters supporting the `file` property, use `subdir` to directly
            // output them in the `dir` directory
            {type: 'cobertura', subdir: '.', file: 'cobertura.txt'},
            {type: 'lcovonly', subdir: '.', file: 'report-lcovonly.txt'},
            {type: 'teamcity', subdir: '.', file: 'teamcity.txt'},
            {type: 'text', subdir: '.', file: 'text.txt'},
            {type: 'text-summary', subdir: '.', file: 'text-summary.txt'}
        ]
    },

    // web server port
    // CLI --port 9876
    port: 9876,

    // cli runner port
    // CLI --runner-port 9100
    runnerPort: 9100,

    // enable / disable colors in the output (reporters and logs)
    // CLI --colors --no-colors
    colors: true,

    // level of logging
    // possible values: LOG_DISABLE || LOG_ERROR || LOG_WARN || LOG_INFO || LOG_DEBUG
    // CLI --log-level debug
    //logLevel: LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: true,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    // CLI --browsers Chrome,Firefox,Safari
    browsers: ['Chrome'],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 5000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: false,

    // report which specs are slower than 500ms
    // CLI --report-slower-than 500
    reportSlowerThan: 500,

    customLaunchers: {
        ChromeNoSandbox: {
            base: 'Chrome',
            flags: ['--no-sandbox']
        }
    }

};

if (process.env.TRAVIS) {
    configuration.browsers = ['ChromeNoSandbox'];
}

module.exports = function (config) {

    config.set(configuration);
};
